main/src/ca/uwaterloo/flix/language/phase/Typer.scala

```diff
@@ -82,21 +88,119 @@ object Typer extends Phase[ResolvedAst.Root, TypedAst.Root] {
   }
 
   /**
-    * Performs type inference and reassembly on all definitions in the given AST root.
+    * Performs type inference and reassembly on all instances in the given AST root.
     *
     * Returns [[Err]] if a definition fails to type check.
     */
-  private def visitDefs(root: ResolvedAst.Root)(implicit flix: Flix): Validation[Map[Symbol.DefnSym, TypedAst.Def], TypeError] = {
+  private def visitInstances(root: ResolvedAst.Root, classes: Map[Symbol.ClassSym, TypedAst.Class])(implicit flix: Flix): Validation[MultiMap[Symbol.ClassSym, TypedAst.Instance], TypeError] = {
+
+    /**
+      * Reassembles a single instance.
+      */
+    def visitInstance(inst: ResolvedAst.Instance): Validation[TypedAst.Instance, TypeError] = inst match {
+      case ResolvedAst.Instance(doc, mod, sym, tpe, tconstrs, defs0, loc) =>
+        for {
+          defs <- Validation.traverse(defs0)(visitInstanceDefn(_, tconstrs, root))
+        } yield TypedAst.Instance(doc, mod, sym, tpe, tconstrs, defs, loc)
+    }
+
+    /**
+      * Checks for overlap of instance types, assuming the instances are of the same class.
+      */
+    def checkOverlap(inst1: TypedAst.Instance, inst2: TypedAst.Instance)(implicit flix: Flix): Validation[Unit, TypeError] = {
+      Unification.unifyTypes(inst1.tpe, inst2.tpe) match {
```

Comment:
Reading THIH I see that they use type schemes here, but we have  a type?

Reply:
Can you point to where you see this? From my reading they use types.

```haskell
data Pred = IsIn Id Type
    deriving Eq

overlap :: Pred → Pred → Bool
overlap p q = defined (mguPred p q)
```

Reply:
You are right :)